internal/pkg/plan/plan.go

```diff
@@ -0,0 +1,103 @@
+package plan
+
+import (
+	"github.com/merico-dev/stream/internal/pkg/config"
+	"github.com/merico-dev/stream/internal/pkg/plugin"
+	"github.com/merico-dev/stream/internal/pkg/statemanager"
+)
+
+// Plan is an "Actions" plan, it includes all changes should be take with plugins.
+type Plan struct {
+	Changes []*Change
+}
+
+// ExecuteResult holds the execution result with a single plugin.
+// The map's key is a plugin's name.
+// The Result is Succeeded when map's value error is nil.
+type ExecuteResult map[string]error
+
+// Execute will execute all changes included in the Plan and put the execution result into a ExecuteResult channel.
+func (p *Plan) Execute(execResultChan chan<- ExecuteResult) {
+	for _, c := range p.Changes {
+		// We will consider how to execute Action concurrently later.
+		// It involves dependency management.
+		_, err := c.Action(c.Tool)
+		execResultChan <- ExecuteResult{c.Tool.Name: err}
+	}
+	close(execResultChan)
+}
+
+// ActionFunc is a function that Do Action with a plugin. like:
+// plugin.Install() / plugin.Reinstall() / plugin.Uninstall()
+type ActionFunc func(tool *config.Tool) (bool, error)
+
+// Change is a wrapper with a single Tool and its Action should be execute.
+type Change struct {
+	Tool       *config.Tool
+	ActionName statemanager.ComponentAction
+	Action     ActionFunc
+}
+
+// MakePlan takes "State Manager" & "Config" then do some calculate and return a Plan.
+// All actions should be execute is included in this Plan.changes.
+func MakePlan(smgr statemanager.Manager, cfg *config.Config) *Plan {
+	if cfg == nil {
+		return nil
+	}
+	goThroughConfig := func(plan *Plan, states statemanager.States, cfg *config.Config) {
+		// This for-loop is to filter all the Tools in cfg that need some actions
+		for _, tool := range cfg.Tools {
+			state := smgr.GetState(tool.Name)
+			// The tool has state record at States, so it need to be Installed or Reinstalled if at "not ready" status
+			if state != nil {
+				delete(states, tool.Name)
+				switch state.Status {
+				case statemanager.StatusUninstalled:
+					plan.Changes = append(plan.Changes, &Change{
+						Tool:       &tool,
+						ActionName: statemanager.ActionInstall,
+						Action:     plugin.Install,
+					})
+				case statemanager.StatusFailed:
+					plan.Changes = append(plan.Changes, &Change{
+						Tool:       &tool,
+						ActionName: statemanager.ActionReinstall,
+						Action:     plugin.Reinstall,
+					})
+				}
+				// The tool hasn't state record at States, so it need to be Installed
```

Comment:
Comment should be on the same indentation level as the code.

Reply:
The indentation level is decided by the IDE, I've tried modify the indentation level and it will be revert to this style after "format"

Reply:
Let me do some tests and try to figure it out. Thanks for the input.